### PR TITLE
BAEL-3352 - Websockets with the Play Framework and Akka

### PR DESCRIPTION
**Pull request by Alfred-Samanga**

### CODE REVIEW

File: play-framework/websockets/app/actors/Messenger.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+import java.time.OffsetDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.UUID;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Created by alfred on 02 Mar 2019
+ */
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/actors/Messenger.java]: Let's remove the author comments here, please.

File: play-framework/websockets/app/actors/Messenger.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Override
+    public void postStop() throws Exception {
+        log.info("Messenger actor stopped at {}",
+          OffsetDateTime.now().format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
+    }
+
+    private void onSendMessage(JsonNode jsonNode) {
+        final RequestDTO requestDTO = MessageConverter.jsonNodeToRequest(jsonNode);
+        final String message = requestDTO.getMessage().toLowerCase();
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/actors/Messenger.java]: Let's remove the `final`s. They're not adding anything.

File: play-framework/websockets/app/actors/Messenger.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    private CompletionStage<HttpResponse> getRandomMessage() {
+        int postId = ThreadLocalRandom.current().nextInt(0, 100);
+        return Http.get(getContext().getSystem()).singleRequest(
+          HttpRequest.create("https://jsonplaceholder.typicode.com/posts/" + postId)
+        );
+    }
+
+    private void discardEntity(HttpResponse httpResponse, Materializer materializer) {
+        final HttpMessage.DiscardedEntity discarded = httpResponse.discardEntityBytes(materializer);
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/actors/Messenger.java]: Do we need the `discarded` temp variable ? Can we not just call fluently from `httpResponse.discardEntityBytes(...)`?

File: play-framework/websockets/app/controllers/HomeController.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Inject
+    public HomeController(ActorSystem actorSystem, Materializer materializer) {
+        this.actorSystem = actorSystem;
+        this.materializer = materializer;
+    }
+
+    public Result index(Http.Request request) {
+        String url = routes.HomeController.socket().webSocketURL(request);
+        //String url = routes.HomeController.akkaStreamsSocket().webSocketURL(request);
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/controllers/HomeController.java]: Why is this commented out? If it's for the reader to experiment with a different option, then comment to them the reason for it being commented out.

File: play-framework/websockets/app/controllers/HomeController.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    private Flow<JsonNode, JsonNode, ?> createFlowForActor() {
+        return ActorFlow.actorRef(out -> Messenger.props(out), actorSystem, materializer);
+    }
+
+    public WebSocket akkaStreamsSocket() {
+        return WebSocket.Json.accept(
+          request -> {
+              Sink<JsonNode, ?> in = Sink.foreach(System.out::println);
+              final MessageDTO messageDTO = new MessageDTO("1", "1", "Title", "Test Body");
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/controllers/HomeController.java]: Extraneous `final`

File: play-framework/websockets/app/dto/MessageDTO.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
+package dto;
+
+/**
+ * Created by alfred on 07 October 2019
+ */
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/dto/MessageDTO.java]: Author comment to be removed

File: play-framework/websockets/app/dto/RequestDTO.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package dto;
+
+/**
+ * Created by alfred on 07 October 2019
+ */
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/dto/RequestDTO.java]: Author comment

File: play-framework/websockets/app/utils/MessageConverter.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
+package utils;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import dto.MessageDTO;
+import dto.RequestDTO;
+
+/**
+ * Created by alfred on 07 Jun 2019
+ */
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/utils/MessageConverter.java]: Author comment

File: play-framework/websockets/app/utils/MessageConverter.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+public class MessageConverter {
+    public static MessageDTO jsonNodeToMessage(JsonNode jsonNode) {
+        ObjectMapper mapper = new ObjectMapper();
+        final MessageDTO messageDTO = mapper.convertValue(jsonNode, MessageDTO.class);
+        return messageDTO;
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/utils/MessageConverter.java]: no need for the temp variable - just return `mapper.convertValue(...)`'s result

File: play-framework/websockets/app/utils/MessageConverter.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    public static JsonNode messageToJsonNode(MessageDTO messageDTO) {
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode jsonData = mapper.convertValue(messageDTO, JsonNode.class);
+        return jsonData;
+    }
+    public static RequestDTO jsonNodeToRequest(JsonNode jsonNode) {
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (ashleyfrieze) [inline play-framework/websockets/app/utils/MessageConverter.java]: FWIW this is not the ideal way to use `ObjectMapper`. You don't have to change it, but it's common not to keep creating new instances, and use a `private static final ObjectMapper OBJECT_MAPPER` constant

File: play-framework/websockets/app/actors/Messenger.java (hunk 1 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/actors/Messenger.java]: Okay sure. I have removed these everywhere they occurred.

File: play-framework/websockets/app/actors/Messenger.java (hunk 3 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/actors/Messenger.java]: Fixed.

File: play-framework/websockets/app/controllers/HomeController.java (hunk 4 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/controllers/HomeController.java]: It was to give the user the ability to switch. So I have added the comments

File: play-framework/websockets/app/controllers/HomeController.java (hunk 5 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/controllers/HomeController.java]: I have removed all redundant final keywords

File: play-framework/websockets/app/dto/MessageDTO.java (hunk 6 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/dto/MessageDTO.java]: Removed.

File: play-framework/websockets/app/utils/MessageConverter.java (hunk 9 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/utils/MessageConverter.java]: Done.

File: play-framework/websockets/app/utils/MessageConverter.java (hunk 10 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/utils/MessageConverter.java]: I have changed to use the private static final constant.

File: play-framework/websockets/app/dto/RequestDTO.java (hunk 7 above)
Comment (Alfred-Samanga) [reply play-framework/websockets/app/dto/RequestDTO.java]: Removed comment.

**Pull request merged**